### PR TITLE
Request for Placement and Match Request Withdrawal Emails

### PR DESCRIPTION
**Pull request by davidatkinsuk**
Relates to https://dsdmoj.atlassian.net/browse/APS-163

See the most recent comment on APS-163 for an explanation of when emails will be sent

This PR also includes some improvements to the existing Withdrawal Cascade integration tests

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1BookingEmailService.kt [hunk 1]
```diff
@@ -80,9 +80,9 @@ class Cas1BookingEmailService(
     }
 
     val area = application.apArea
-    area?.emailAddress?.let { email ->
+    area?.emailAddress?.let { cruEmail ->
```
Comment (edavey) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1BookingEmailService.kt:83]: 👍

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/WithdrawalTest.kt [hunk 2]
```diff
@@ -505,30 +506,70 @@ class WithdrawalTest : IntegrationTestBase() {
...
+     * --------------------------------------------------------------------------------------------------
+     * application                       | YES       | YES      | NO       | NO        | NO             |
+     *  - assessment                     | YES       | NO       | NO       | NO        | NO             |
+     *  - request for placement 1        | YES       | YES      | NO       | NO        | NO             |
+     *    - match request 1              | YES       | NO       | NO       | NO        | NO             |
+     *      - booking 1 arrival pending  | YES       | YES      | YES      | YES       | NO             |
+     *    - match request 2              | YES       | NO       | NO       | YES       | NO             |
+     *  - request for placement 2        | YES       | YES      | NO       | NO        | YES            |
+     *  - match request 3                | YES       | NO       | NO       | NO        | NO             |
+     *    - booking 2 has arrival        | NO        | NO       | NO       | NO        | NO             |
```
Comment (edavey) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/WithdrawalTest.kt]: this is a thing of beauty, thks

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/notification/EmailNotificationAsserter.kt [hunk 3]
```diff
@@ -11,10 +13,13 @@ import kotlin.math.exp
 @Component
 class EmailNotificationAsserter {
 
+  var log: Logger = LoggerFactory.getLogger(this::class.java)
+
   private val requestedEmails = mutableListOf<EmailRequest>()
 
   @EventListener
   fun consumeEmailRequestedEvent(emailRequested: SendEmailRequestedEvent) {
+    log.info("Email requested ${emailRequested.request}")
```
Comment (edavey) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/notification/EmailNotificationAsserter.kt:22]: handy, thks

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/PlacementRequestsTest.kt [hunk 4]
```diff
@@ -1620,7 +1620,7 @@ class PlacementRequestsTest : IntegrationTestBase() {
     }
 
     @Test
-    fun `Withdraw Placement Request returns 200, sets isWithdrawn to true and sends email to CRU when no booking`() {
+    fun `Withdraw Placement Request returns 200, sets isWithdrawn to true and sends email to CRU when no booking and applicant`() {
```
Comment (edavey) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/PlacementRequestsTest.kt:1623]: should we try to be a bit more explicit about the scenario here, e.g. "...when their is no booking and an applicant exists as the PR was automatically created on submission of an application with known release date". A bit verbose but might be helpful documentation?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/cas1/Cas1PlacementRequestEmailServiceTest.kt [hunk 5]
```diff
@@ -71,21 +74,95 @@ class Cas1PlacementRequestEmailServiceTest {
...
+      hasPlacementApplication = false,
+    )
+
+    service.placementRequestWithdrawn(placementRequest)
+
+    mockEmailNotificationService.assertNoEmailsRequested()
+  }
+
+  @Test
+  fun `placementRequestWithdrawn does not send email to applicant if placement request linked to placement application `() {
```
Comment (edavey) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/cas1/Cas1PlacementRequestEmailServiceTest.kt:109]: again would it be useful to explain how this comes to be? e.g. "... when automatically generated on submission of an application with a known release date"?

**Review by edavey — APPROVED** [review]
tip top, many thanks for making this so clear and thorough.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/PlacementRequestsTest.kt (hunk 4 above)
Comment (davidatkinsuk) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/PlacementRequestsTest.kt]: Good idea. I've had a go at rewording. I'm hoping over the next few weeks we can keep refactoring to the point where this logic makes more sense and doesn't necessarily need so much explanation (but at the moment I completely agree it does!)

**Pull request merged**